Add Pressure Stall Information (PSI) to Systemmonitor integration

ajfriesen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

I want all Home Assistant users to be able to check if their system is bound by any particular resource.
Especially memory, but also IO and CPU.
Used memory is a very bad metric to make decisions on.
The system could use more than 99% of memory and could be totally fine or overloaded.
Memory is often shared, then there is cache, buffers,etc.
Used memory is good memory, and makes it hard to use it as a constraining measure.

Then there is OOM, but then it is too late.

With PSI for memory, we can check if the system was or is about to be bottlenecked by memory, cpu or IO.

Pressure Stall Information (PSI) is a relatively new metric from the Linux Kernel.
I can be found at `/proc/pressure/{cpu,memory,io}`
Example:
```
some avg10=0.00 avg60=0.00 avg300=0.00 total=376
full avg10=0.00 avg60=0.00 avg300=0.00 total=85
```

The `some` statistic gives the percentage of time some (one or more) tasks were delayed due to a lack of resources, for example, a lack of memory.

The `full` metric indicates the percentage of time in which all tasks were delayed by lack of resources, i.e., the amount of completely unproductive time.
A high full number indicates a loss of overall throughput – the total amount of work done decreases due to lack of resources.

The `total` absolute stall time (in us) is tracked and exported as well, to allow detection of latency spikes which wouldn’t necessarily make a dent in the time averages, or to average trends over custom time frames.

Sources:

- https://facebookmicrosites.github.io/psi/docs/overview
- https://www.kernel.org/doc/html/v6.12/accounting/psi.html

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
  - https://github.com/home-assistant/operating-system/pull/4279
- Link to documentation pull request: TBD
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
  - I had to build the my own HA OS, since PSI was just merged into dev. But could run my changes as a custom component and did get all pressure values. I tested by installing the esphome builder and built a few configs and could see cpu, memory and io pressure values go up here and there. 
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

https://github.com/home-assistant/home-assistant.io/pull/40798

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ajfriesen committed:
Add psi psi_parser

Add a parser to get pressure stall information
from the Linux kernel in /proc/pressure/{cpu,memory,io}

ajfriesen committed:
Add psi metrics as sensor

ajfriesen committed:
Add tests for psi metrics

ajfriesen committed:
Add missing snapshots for psi tests

ajfriesen committed:
Formatting

ajfriesen committed:
Add missing types for mypy

ajfriesen commented:
@gjohansson-ST 

As the code owner of this repo, could you have a look at this?

gjohansson-ST commented on homeassistant/components/systemmonitor/psi_parser.py:
```diff
@@ -0,0 +1,65 @@
+                key, value = item.split("=")
+                # Convert values to float, except for 'total' which is an integer
+                if key == "total":
+                    values[key] = int(value)
```
I think we need to guard if the values are not convertable (for whatever reason)?

ajfriesen replied:
I added a try-except block around this.
Make sense to catch it if ever this changes or is malformed (unlikely though).

gjohansson-ST commented on homeassistant/components/systemmonitor/psi_parser.py:
```diff
@@ -0,0 +1,65 @@
+    try:
+        with open(file_path) as f:
+            content = f.read()
+    except FileNotFoundError:
```
Do we need an exception for not being authorized to read the file?

gjohansson-ST commented on tests/components/systemmonitor/test_psi_sensor.py:
```diff
@@ -0,0 +1,92 @@
+"""Test System Monitor PSI sensor."""
```
Include additional tests in the already existing `test_sensor.py` file, don't create another test file.

gjohansson-ST commented on tests/components/systemmonitor/test_psi_sensor.py:
```diff
@@ -0,0 +1,92 @@
+
+from tests.common import MockConfigEntry
+
+MOCK_PRESSURE_INFO = {
```
Add mock info in `conftest.py` to the existing mock and these sensors should be already covered by the existing tests.

gjohansson-ST commented:
Please take a look at the failing CI and run `pre-commit` before pushing your commits to ensure it passes 👍

gjohansson-ST commented on homeassistant/components/systemmonitor/psi_parser.py:
```diff
@@ -0,0 +1,65 @@
+"""Parses Pressure Stall Information (PSI) from /proc/pressure files on Linux systems."""
```
Put this in the existing `util.py`

ajfriesen committed:
Move content of psi_parser.py to util.py

ajfriesen committed:
Remove check for FileNotFoundError

File not found will be catched by OSError as well.
No need for checking twice.

ajfriesen committed:
Add try except block when splitting /proc/pressure

ajfriesen committed:
Move test_psi_sensor.py into test_sensor.py

ajfriesen committed:
Increase test cases and add newline

ajfriesen committed:
Update snapshot for systemmonitor tests

ajfriesen committed:
Add encoding for reading /proc/pressure

ajfriesen committed:
Changes through linter

ajfriesen commented:
rebased up the dev branch because there were some conflicts

gjohansson-ST commented:
Needs a rebase because of conflicts, put it back for review when fixed 👍

ajfriesen committed:
Remove not needed separate psi sensor test

ajfriesen committed:
Adjust test counts

ajfriesen committed:
Fix whitespace

joostlek commented on homeassistant/components/systemmonitor/sensor.py:
```diff
@@ -459,6 +464,226 @@ class SysMonitorSensorEntityDescription(SensorEntityDescription):
+        icon="mdi:gauge",
+        value_fn=lambda entity: entity.coordinator.data.pressure.get("memory", {})
+        .get("some", {})
+        .get("avg10"),
```
Please wrap in parenthesis as it now spans multiple lines

ajfriesen committed:
Switch PSI tests from snapshots to static tests

ajfriesen committed:
Refactor System Monitor PSI tests and fix snapshot collisions

- Replace snapshot assertions with explicit state assertions for PSI sensors
- Exclude PSI sensors from general sensor snapshot tests
- Use entity_id for snapshot naming to resolve fan sensor collisions
- Include full CPU pressure in mock data

ajfriesen committed:
Switch to sentence case for entity names

ajfriesen committed:
Wrap multi-line functions in parenthesis

ajfriesen committed:
Consolidate System Monitor PSI tests into test_sensor

- Move explicit PSI state assertions into the main test_sensor function
- Remove the redundant test_psi_sensor function
- Fix indentation in test_process_sensor_not_loaded

ajfriesen committed:
Use icon translations for PSI sensors

frenck committed:
Merge branch 'dev' into psi

joostlek committed:
Fix
